src/charm.py

```diff
@@ -0,0 +1,208 @@
+                time.sleep(sleeptime)
+        else:
+            self.unit.status = MaintenanceStatus(
+                "Some kubernetes resources missing/not ready"
+            )
+            return
+
+    def _update_status(self, event):
+        self.logger.info("Comparing current state to desired state")
+
+        running = self._check_deployed_resources()
+        if running is True:
+            self.logger.info("Resources are ok.  Unit in ActiveStatus")
+            self.unit.status = ActiveStatus()
+            return
+        else:
+            self.logger.info(
+                "Resources are missing.  Triggering install to reconcile resources"
+            )
+            self.unit.status = MaintenanceStatus(
+                "Missing kubernetes resources detected - reinstalling"
+            )
+            self._install(event)
+            return
+
+    def _remove(self, _):
+        """Remove charm"""
+        # TODO: How should we test this in our integration tests?
+        # Should I set a status or does Juju set one?
+        self.logger.info("Removing kubernetes objects")
+
+        _, manifests_str = self._render_manifests()
+        completed_process = subprocess.run(
+            ["./kubectl", "delete", "-f-"],
+            input=manifests_str.encode("utf-8"),
+            stderr=subprocess.STDOUT,
+            stdout=subprocess.PIPE,
+        )
+        if completed_process.returncode == 0:
+            self.logger.info("Kubernetes objects removed using kubectl")
+        else:
+            self.logger.error(
+                f"Unable to remove kubernetes objects - there may be orphaned resources."
+                f"  kubectl exited with '{completed_process.stdout}'"
+            )
+
+    def _render_manifests(self) -> (list, str):
+        # Load and render all templates
+        self.logger.info(f"Rendering templates from {self.manifest_file_root}")
+        jinja_env = Environment(loader=FileSystemLoader(self.manifest_file_root))
+        manifests = []
+        for f in self._get_manifest_files():
+            self.logger.info(f"Rendering template {f}")
+            f = Path(f).relative_to(self.manifest_file_root)
+            template = jinja_env.get_template(str(f))
+            manifests.append(
+                template.render(
+                    namespace=self.model.name,
+                    image="metacontroller/metacontroller:v0.3.0",
+                )
+            )
+
+        # Combine templates into a single string
+        manifests_str = "\n---\n".join(manifests)
+
+        logging.info(f"rendered manifests_str = {manifests_str}")
+
+        return manifests, manifests_str
+
+    def _check_deployed_resources(self, manifests=None):
+        """Check the status of all deployed resources, returning True if ok"""
+        # TODO: Add checks for other CRDs/services/etc
+        # TODO: ideally check all resources automatically based on the manifest
```

Comment:
I think this TODO encapsulates the hardest bit about methods like this. We really shouldn't merge without a fully-fleshed out version of this.

Backing up a bit, this should also generally be the domain of Juju/Operator Framework. Charm authors shouldn't have to write stuff like the `max_attempts = 20` looping above. Charm authors should be able to say something like this:

```python
def install(self, event):
    ops.please_ensure_exists(
        './src/manifests/*.yaml',
        template_vars={'namespace': self.model.name}
    )
```

The operator framework then instructs Juju "here's some YAML templates and the values to use as variables to use when rendering them". Juju then goes off, renders the YAML, ensures that the YAML exists, and then periodically checks the existing resources to ensure that they haven't drifted from what the charm expects.

To preempt the inevitable "but then we're no better than kubectl": kubectl doesn't have any notion of relations, or config, or lots of other powerful Juju features. Our value add here is that `juju relate charm-a charm-b` or `juju config my-charm foo=bar` can drastically change the YAML that a charm tells Juju to apply. That's something that kubectl will never do. We need to stop being afraid of applying Kubernetes YAML.

Put another way, you're trying to write code here that should belong over in https://github.com/canonical/operator and https://github.com/juju/juju/ :smile: If nothing else, for the simple fact that we should not have to copy/paste this code into every charm ever written.

cc @jnsgruk @petevg @jameinel

Reply:
I don't know where I stand on this, but I could see a few interesting scenarios
1. If I could do something like `juju.create_and_tell_me_if_it_breaks(yaml, event_if_it_breaks)` that would be fantastic.  Juju can establish all the watches etc and should something break (`status!=ready`, object gets deleted, etc) an event of my choosing fires and I can act on it right away.  That feels powerful without juju needing to know too much about the charm implementation details.  This feels like an extension of watching k8s objects into the juju space.  The `-operator` code can create the objects and subscribe to their changes, and the charm code can ask "if you see a change, fire this event"
2. If we're not subscribing to the watches, then I think we want a generic `tell_me_if_these_objects_are_ok(yaml)` that knows how to check the status of various objects (does CRD exist, does deployment have correct number of replicas in ready state, etc).  Agreed that shouldn't be cut/paste between all charms, but I don't think that is `juju` specific code.  That feels generic (could be in a helper library or something like `lightkube`)

If we're passing `juju` the yaml to execute, I don't personally see why we pass unrendered yaml + variables.  The variables depend on logic that the charm knows and juju doesn't, so to me that feels rendering the yaml is the charm's job.  Charm can pass rendered yaml to juju, and juju can execute/watch it.

Reply:
I think these scenarios will/should be covered by the templates/scriptlets work. Let's give the juju team some time to come up with a design for that and provide feedback at that point. Us speculating on different ways how the juju team should implement this isn't very productive.

Reply:
> If we're passing juju the yaml to execute, I don't personally see why we pass unrendered yaml + variables

@ca-scribner Good point, the charm should just pass `ops` or `juju` rendered YAML that needs to exist.

Reply:
Think we've tabled the broader discussion, but also the method is now fleshed out